python/cudf/cudf/core/frame.py

```diff
@@ -3581,64 +3582,148 @@ def _reindex(
         return self._mimic_inplace(result, inplace=inplace)
 
 
-def _get_replacement_values(to_replace, replacement, col_name, column):
+def _get_replacement_values_for_columns(
+    to_replace: Any, value: Any, columns: Dict[Any, Any]
+) -> Tuple[Dict[Any, Any], Dict[Any, Any], Dict[Any, Any]]:
+    """
+    Returns a per column mapping for the values to be replaced, new
+    values to be replaced with and if all the values are empty.
 
-    all_nan = False
-    if (
-        is_dict_like(to_replace)
-        and not isinstance(to_replace, cudf.Series)
-        and replacement is None
+    Parameters
+    ----------
+    to_replace : numeric, str, list-like or dict
+        Contains the values to be replaced.
+    value : numeric, str, list-like, or dict
+        Contains the values to replace `to_replace` with.
+    columns : dict
+        A column to dtype mapping representing dtype of columns.
+
+    Returns
+    -------
+    all_nan_columns : dict
+        A dict mapping of all columns if they contain all na values
+    to_replace_columns : dict
+        A dict mapping of all columns and the existing values that
+        have to be replaced.
+    values_columns : dict
+        A dict mapping of all columns and the corresponding values
+        to be replaced with.
+    """
+    to_replace_columns: Dict[Any, Any] = {}
+    values_columns: Dict[Any, Any] = {}
+    all_nan_columns: Dict[Any, Any] = {}
+
+    if is_scalar(to_replace) and is_scalar(value):
+        to_replace_columns = {col: [to_replace] for col in columns}
+        values_columns = {col: [value] for col in columns}
+    elif cudf.utils.dtypes.is_list_like(to_replace) or isinstance(
+        to_replace, cudf.core.column.ColumnBase
     ):
-        replacement = list(to_replace.values())
-        to_replace = list(to_replace.keys())
-    elif not is_scalar(to_replace):
-        if is_scalar(replacement):
-            all_nan = replacement is None
-            if all_nan:
-                replacement = [replacement] * len(to_replace)
-            # Do not broadcast numeric dtypes
-            elif pd.api.types.is_numeric_dtype(column.dtype):
-                if len(to_replace) > 0:
-                    replacement = [replacement]
-                else:
-                    # If to_replace is empty, replacement has to be empty.
-                    replacement = []
-            else:
-                replacement = utils.scalar_broadcast_to(
-                    replacement,
-                    (len(to_replace),),
-                    np.dtype(type(replacement)),
-                )
-        else:
-            # If both are non-scalar
-            if len(to_replace) != len(replacement):
+        if is_scalar(value):
+            to_replace_columns = {col: to_replace for col in columns}
+            values_columns = {
+                col: [value]
+                if pd.api.types.is_numeric_dtype(columns[col])
+                else [value] * len(to_replace)
```

Comment:
Do we support `to_replace` being a GPU backed object here? Do we suspect it would ever be large?

Reply:
> Do we support `to_replace` being a GPU backed object here? Do we suspect it would ever be large?

Yes, and Yes. `to_replace` can get potentially super large hence changed this to broadcast the scalar to a column.